### PR TITLE
fix(sampler): logprobs overflow due to non-deterministic sort when using top_p sampling

### LINKED ISSUE #4876: [Bug] Logprobs overflow to -3.4e+38
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

Image

logprobs overflow to the maximum negative value of fp32

### Reproduction

I'm using Qwen2.5-14B-Instruct

command:

```python
sampling_params = {
    "temperature": 0.9,
    "top_p": 0.9,
    "skip_special_tokens": False,
    "stop": "<|im_end|>",
}

ret = await self.engine.async_generate(
    input_ids=ids,
    sampling_params=sampling_params,
    return_logprob=True,
)
```

### Environment

```
Python: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H800
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.3, V12.3.107
CUDA Driver Version: 535.161.08
PyTorch: 2.6.0+cu124
sglang: 0.4.4.post1
sgl_kernel: 0.0.5.post3
flashinfer: 0.2.3+cu124torch2.5
triton: 3.2.0
transformers: 4.48.3
torchao: 0.9.0
numpy: 1.26.4
aiohttp: 3.9.1
fastapi: 0.115.5
hf_transfer: 0.1.9
huggingface_hub: 0.26.2
interegular: 0.3.3
modelscope: 1.23.1
orjson: 3.10.11
packaging: 23.2
psutil: 5.9.4
pydantic: 2.10.5
multipart: 0.0.18
zmq: 25.1.2
uvicorn: 0.22.0
uvloop: 0.21.0
vllm: 0.7.3.dev68+g9cf47594.d20250213
openai: 1.59.6
anthropic: 0.49.0
decord: 0.6.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    CPU Affinity    NUMA Affinity GPU NUMA ID
GPU0     X      NV8     NV8     NV8     NV8     NV8     NV8     NV8     PIX     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     0,8,16,24,34    0    N/A
GPU1    NV8      X      NV8     NV8     NV8     NV8     NV8     NV8     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     2,10,18,30      2    N/A
GPU2    NV8     NV8      X      NV8     NV8     NV8     NV8     NV8     SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     6,14,22,28      3    N/A
GPU3    NV8     NV8     NV8      X      NV8     NV8     NV8     NV8     SYS     SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     4,12,20,26      1    N/A
GPU4    NV8     NV8     NV8     NV8      X      NV8     NV8     NV8     SYS     SYS     SYS     SYS     SYS     PIX     SYS     SYS     SYS     1,9,19,27,33    4    N/A
GPU5    NV8     NV8     NV8     NV8     NV8      X      NV8     NV8     SYS     SYS     SYS     SYS     SYS     SYS     PIX     SYS     SYS     3,11,15,21      6    N/A
GPU6    NV8     NV8     NV8     NV8     NV8     NV8      X      NV8     SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX     SYS     7,25,31,39      7    N/A
GPU7    NV8     NV8     NV8     NV8     NV8     NV8     NV8      X      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX     5,13,17,23      5    N/A
NIC0    PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC1    PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX      X      SYS     SYS     SYS     SYS     SYS     SYS     SYS
NIC2    SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS     SYS     SYS     SYS     SYS     SYS
NIC3    SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS     SYS     SYS     SYS     SYS
NIC4    SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS     SYS     SYS     SYS
NIC5    SYS     SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS     SYS     SYS
NIC6    SYS     SYS     SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS     SYS
NIC7    SYS     SYS     SYS     SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X      SYS
NIC8    SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_200
  NIC1: mlx5_400
  NIC2: mlx5_401
  NIC3: mlx5_402
  NIC4: mlx5_403
  NIC5: mlx5_404
  NIC6: mlx5_405
  NIC7: mlx5_406
  NIC8: mlx5_407


ulimit soft: 1048576
```

### PR DESCRIPTION
**Pull request by zhc7**
## Motivation

fixes #4876

## Modifications

already stated in #4876

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [x] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by zhc7** [comment]
cc @zhaochenyang20  , thanks!

File: python/sglang/srt/layers/sampler.py [hunk 1]
```diff
@@ -238,8 +238,10 @@ def top_p_normalize_probs_torch(
     # See also top_k_top_p_min_p_sampling_from_probs_torch
     probs_sort, probs_idx = probs.sort(dim=-1, descending=True)
     probs_sum = torch.cumsum(probs_sort, dim=-1)
-    probs_sort[(probs_sum - probs_sort) > top_ps.view(-1, 1)] = 0.0
-    probs_sort.div_(probs_sort.sum(dim=-1, keepdim=True))
+    # sort is non-deterministic, prevent probs masked to zero
```
Comment (zhaochenyang20) [inline python/sglang/srt/layers/sampler.py:241]:
```python
# sort is non-deterministic, prevent probs masked to zero
```

This is not clear to me. Could you explain it more.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/layers/sampler.py (hunk 1 above)
Comment (zhc7) [reply python/sglang/srt/layers/sampler.py:241]: sure, the key is that if two probs are the same, they might be sorted in arbitrary order. if this happens at the border of top_p, one of it will be masked to zero.

Comment (zhaochenyang20) [reply python/sglang/srt/layers/sampler.py:241]: could you provide me one code examples?

Comment (zhc7) [reply python/sglang/srt/layers/sampler.py:241]: I can give an example:

```
>>> probs = torch.tensor([[0.85, 0.07, 0.07, 0.01]])
>>> probs_sort, probs_idx = probs.sort(dim=-1, descending=True)
>>> probs_sort, probs_idx
(tensor([[0.8500, 0.0700, 0.0700, 0.0100]]), tensor([[0, 1, 2, 3]]))
```

it is also possible that `probs_idx` is `tensor([[0, 2, 1, 3]]`

so in the final return at L245, the final tensor might be either `tensor([[0.9239, 0.0761, 0.0, 0.0]])` or `tensor([[0.9239, 0.0, 0.0761, 0.0]])`. but token 1 and 2 both might be selected by sampler.

the token selection logic is in this function:

```python
def top_k_top_p_min_p_sampling_from_probs_torch(
    probs: torch.Tensor,
    top_ks: torch.Tensor,
    top_ps: torch.Tensor,
    min_ps: torch.Tensor,
    need_min_p_sampling: bool,
):
    """A top-k, top-p and min-p sampling implementation with native pytorch operations."""
    probs_sort, probs_idx = probs.sort(dim=-1, descending=True)
    probs_sum = torch.cumsum(probs_sort, dim=-1)
    probs_sort[
        torch.arange(0, probs.shape[-1], device=probs.device).view(1, -1)
        >= top_ks.view(-1, 1)
    ] = 0.0
    probs_sort[(probs_sum - probs_sort) > top_ps.view(-1, 1)] = 0.0

    if need_min_p_sampling:
        min_p_thresholds = probs_sort[:, 0] * min_ps
        probs_sort[probs_sort < min_p_thresholds.view(-1, 1)] = 0.0

    sampled_index = torch.multinomial(probs_sort, num_samples=1)
    # int32 range is enough to represent the token ids
    probs_idx = probs_idx.to(torch.int32)
    batch_next_token_ids = torch.gather(probs_idx, dim=1, index=sampled_index).view(-1)
    return batch_next_token_ids
```

which also relies on a `torch.sort`. if two `torch.sort` gives different order regarding same value, and that border token happened to be selected, we might get log0, which is -3.4e+38 in torch.

Comment (zhaochenyang20) [reply python/sglang/srt/layers/sampler.py:241]: Could you just give me a python scripts to run, like `python test.py`.

Comment (zhc7) [reply python/sglang/srt/layers/sampler.py:241]: this is rather hard to provide... as it will require a reproduction of the internal non-determinism of pytorch sorting algo, and I'm not sure how to reproduce that. the non-deterministic behavior is observed in a rollout process, maybe 1 out of 10000 tokens results in an 3.4e+38 logprob. I'm not able to reproduce this with a code snippet.

Comment (zhc7) [reply python/sglang/srt/layers/sampler.py:241]: the settings to reproduce the issue is provided in the related issue: https://github.com/sgl-project/sglang/issues/4876, this might be helpful. it's very hard to stably reproduce this without large scale rollout 😢
